packages/flutter/lib/src/widgets/app.dart

```diff
@@ -690,35 +742,147 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
+  /// when `de` is not supported and `zh` is listed before `fr` (German is closer to French
+  /// than Chinese).
+  static Locale defaultLocaleListResolution(List<Locale> preferredLocales, Iterable<Locale> supportedLocales) {
+    // preferredLocales can be null when called before the platform has had a chance to
+    // initialize the locales. Platforms without locale passing support will provide an empty list.
+    // We default to the first supported locale in these cases.
+    if (preferredLocales == null || preferredLocales.isEmpty) {
       return supportedLocales.first;
     }
+    // Hash the supported locales because apps can support many locales and would
+    // be expensive to search through them many times.
+    final Map<int, Locale> allSupportedLocales = HashMap<int, Locale>();
+    final Map<int, Locale> languageAndCountryLocales = HashMap<int, Locale>();
+    final Map<int, Locale> languageAndScriptLocales = HashMap<int, Locale>();
+    final Map<int, Locale> languageLocales = HashMap<int, Locale>();
+    final Map<int, Locale> countryLocales = HashMap<int, Locale>();
+    for (Locale locale in supportedLocales) {
+      allSupportedLocales[ui.hashValues(locale.languageCode.hashCode, locale.scriptCode.hashCode, locale.countryCode.hashCode)] ??= locale;
+      languageAndCountryLocales[ui.hashValues(locale.languageCode.hashCode, locale.countryCode)] ??= locale;
+      languageAndScriptLocales[ui.hashValues(locale.languageCode.hashCode, locale.scriptCode.hashCode)] ??= locale;
+      languageLocales[locale.languageCode.hashCode] ??= locale;
+      countryLocales[locale.countryCode.hashCode] ??= locale;
+    }
 
+    // Since languageCode-only matches are possibly low quality, we don't return
+    // it instantly when we find such a match. We check to see if the next
+    // preferred locale in the list has a high accuracy match, and only return
+    // the languageCode-only match when a higher accuracy match in the next
+    // preferred locale cannot be found.
     Locale matchesLanguageCode;
-    for (Locale locale in supportedLocales) {
-      if (locale == newLocale)
-        return newLocale;
-      if (locale.languageCode == newLocale.languageCode)
-        matchesLanguageCode ??= locale;
+    Locale matchesCountryCode;
+    // Loop over user's preferred locales
+    for (int localeIndex = 0; localeIndex < preferredLocales.length; localeIndex += 1) {
+      final Locale userLocale = preferredLocales[localeIndex];
+      // Look for perfect match.
+      if (allSupportedLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode, userLocale.countryCode.hashCode))) {
+        return userLocale;
+      }
+      // Look for language+script match.
+      if (userLocale.scriptCode != null) {
+        final Locale match = languageAndScriptLocales[ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode)];
+        if (match != null)
```

Comment:
nit: for consistency with the other code in this file: use {} even for one line ifs?

Reply:
Also below.